rfcs/rfc-0000-discriminators.md

```diff
@@ -0,0 +1,123 @@
+- Feature Name: Discriminators
+- Start Date: 06/05/2023
+- RFC PR: https://github.com/revoltchat/rfcs/issues/0000
+- Tracking Issue: https://github.com/revoltchat/revolt/issues/0000
+- Status: draft
+
+# Summary
+
+Revolt should introduce discriminators which serve as part of a semi-hidden extension to
+usernames which allows multiple users to have the same username, the discriminator would
+not be visible in most circumstances except when viewing a person's profile and if you
+were to add said person by their username.
+
+# Motivation
+
+This RFC aims to move forwards [prior discussion on the feature suggestions forum](https://github.com/orgs/revoltchat/discussions/61),
+it stands to solve multiple problems when implemented correctly:
+
+- Users should in most, if not all, cases be able to pick their chosen username.
+- This should curb any incentive to sell / steal accounts for "rare usernames".
+
+  Certain "rare" discriminators can also be prohibited.
+
+- Allow users who do not want to be easily discoverable to stay hidden.
+
+  Just a 4-digit discriminator provides 10,000 unique users.
```

Comment:
if you were to include letters as well as numbers, you could bump this number up to **1,679,616**.

and if you REALLY wanted to, you could even include lowercase characters and get that number as high as **14,776,336**. and that's all just with 4 characters still!

although this *does* open up the possibility of creating bad words in the discriminator... but i can't imagine that'd be too difficult to mitigate if you were to just block a handful of possible combinations. you'd still have millions of discriminators in the end, which should be more than enough for almost anyone who comes to the platform.

Reply:
> although this _does_ open up the possibility of creating bad words in the discriminator... but i can't imagine that'd be too difficult to mitigate if you were to just block a handful of possible combinations. you'd still have millions of discriminators in the end, which should be more than enough for almost anyone who comes to the platform.

Putting a limit of 2 letters could prevent this as well while also still allowing a lot of possible combos

Reply:
with a max of 2 letters, that'd be around **129,600** combinations if my math is correct. and with lower and uppercase, this can be bumped up to about **384,400**.

although even if you limit it to 2 letter characters at a time, it'd still be possible to create unfortunate combinations. a couple that immediately come to mind are `SH17` and `PI55`

it could be easier to moderate than 4 letter characters, but it does mean sacrificing a considerable amount of possible combinations. though even then, it's still a huge improvement over 10,000 with just plain digits.

Reply:
why would it be more challenging to use the full range of numbers and letters? 
0-9 + A-F is just base 16.
0-9 + A-Z would be base 36.

converting to/from base 36 is incredibly simple. JavaScript even has built-in functions that can do this for you:
- https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/parseInt
- https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/toString

but even if you don't have this luxury, it's still fairly easy to write a couple functions to do these conversions. I myself had to do it for base64, and it was only about ~10 lines of code for each.

and if you're worried about hitting integer limits, most languages can handle unsigned 32 bit integers, which gives you a maximum number size of **4,294,967,295**... which, needless to say, is *far* beyond the number of combinations we're possibly looking at here.

as for the other points you brought up... well, i can't really argue much against them. being proactive is the best way to prevent bad discriminators from getting assigned to people, but that requires additional work, and of course you can't guarantee you'll catch everything.

although, memorizing discriminators is kind of a problem regardless imo. in my mind, if they're gonna be difficult to memorize anyway, you might as well make them more functionally useful (and therefore, increase the number of unique combos).

Reply:
I've added a table comparing (in my opinion) the relative advantages and disadvantages of different discriminator formats: https://github.com/insertish/revolt-rfcs/blob/discriminators/rfcs/rfc-0000-discriminators.md#unresolved-questions

I'm personally most interested in base-16 discriminators.

Reply:
Currently 4-digit is most appealing for the size of the platform, and this may be trivially expanded in the future.
Wrote this into the reference-level explanation.